lib/galaxy/web/framework/decorators.py

```diff
@@ -84,7 +84,7 @@ def decorator(self, trans, *args, **kwargs):
             else:
                 return trans.show_error_message(
                     'You must be <a target="galaxy_main" href="%s">logged in</a> to %s.'
-                    % (url_for(controller='user', action='login', webapp=webapp), verb), use_panels=use_panels)
+                    % (url_for(controller='login', webapp=webapp), verb), use_panels=use_panels)
```

Comment:
What do you think about something like eb1666d92201c6205d176bbdfe1eca790b4ab1ad ?

Reply:
Thanks for looking into it. It's better with its own controller. We spent a lot of time in the past to untangle Galaxy from the Toolshed by avoiding the usage of `webapp` which should be deprecated to ensure that the two do not mix again.

Reply:
I don't know about the double redirect there though, that seems bad as well ... it's also just a detail that could be fixed easily with a TS-specific decoator. The point is more that that suggestion fixes https://github.com/galaxyproject/galaxy/issues/12735 completely.

Reply:
Also my understanding is that it's fine and desirable that the toolshed cleanly re-uses Galaxy code, the problem is with Galaxy using TS code.

Reply:
Its not cleanly reused if there are `webapps` based conditionals in Galaxy code. This was actually a huge challenge when untangling the two, the concept of distinguishing code paths between the two using `webapps` lead to a plethora of issue when we rewrote the login and registration managers in Galaxy. I am fine with any change as long as it is confined to the Toolshed.